pulp_rpm/app/modulemd.py

```diff
@@ -2,14 +2,50 @@
 import os
 import tempfile
 
-from pulpcore.app.models.content import Artifact
+from pulpcore.app.models.content import Artifact, Content
 from pulp_rpm.app.constants import PULP_MODULE_ATTR, PULP_MODULEDEFAULTS_ATTR
 
 import gi
 gi.require_version('Modulemd', '2.0')
 from gi.repository import Modulemd as mmdlib  # noqa: E402
 
 
+def resolve_module_packages(version):
+    """
+    Decide which packages to add/remove based on modular data.
+
+    Args:
+        version (pulpcore.app.models.RepositoryVersion): current incomplete repository version
+    """
+    repo = version.repository
+    base_version = version.base_version
+    if base_version:
+        previous_version = base_version
+    else:
+        previous_version = repo.versions.get(number=repo.last_version - 1)
+
+    previous_modules = previous_version.content.filter(pulp_type="rpm.modulemd")
+    current_modules = version.content.filter(pulp_type="rpm.modulemd")
+
+    added_modules = current_modules.difference(previous_modules)
+    removed_modules = previous_modules.difference(current_modules)
+
+    added_module_package_ids = set(added_modules.values_list("rpm_modulemd__packages__pk",
+                                   flat=True))
+    removed_module_package_ids = set(removed_modules.values_list("rpm_modulemd__packages__pk",
+                                     flat=True))
+    current_module_package_ids = set(current_modules.values_list("rpm_modulemd__packages__pk",
+                                     flat=True))
+    current_package_ids = set(version.content.filter(pulp_type="rpm.package").values_list("pk",
+                              flat=True))
+
+    package_ids_to_remove = removed_module_package_ids.difference(current_module_package_ids)
+    package_ids_to_add = added_module_package_ids.difference(current_package_ids)
+
+    version.add_content(Content.objects.filter(pk__in=package_ids_to_add))
+    version.remove_content(Content.objects.filter(pk__in=package_ids_to_remove))
```

Comment:
I tested this with the following setup and the logic seems to be correct
```
Version 1
=========
module A
*foo
*bar
*beta

module B
*bar
*alpha


Version 2   ( add module C, remove module A)
=========
Module B ( no change)

+ module C
*alpha
*mega
*beta

-Module A

=========
In [3]: package_ids_to_remove                                                                                                                                                                                      
Out[3]: {'foo'}

In [4]: package_ids_to_add                                                                                                                                                                                         
Out[4]: {'mega'}
```

Important to note that package `bar` does not get removed because it is referenced in Module B that stays in the repo.